### DISCUSSION TITLE
Dynamic CSP rules

### DISCUSSION THREAD

**Discussion by MrCasquette** [directus/directus#23854] [Feature Requests]
### Describe the Improvement

It's maybe impossible. But 'impossible is not Directus' Right? 😅
As mentionned [here](https://github.com/directus/directus/issues/21937)

It's hard to figure out how to add CSP rules for live preview functionality.
But in addition to that, I'll add that it's a mess if you're managing multiple blogs, as in a multi-tenant application.
I had to manually write in the docker compose file, stop and start the docker and wait for it to work properly...

Is there any way to add it dynamically everytime users add a tenant / blog or maybe change this rule?

Thank you Rabbits you rocks 🐇 🤘

**Comment by br41nslug** [comment]
> It's maybe impossible. But 'impossible is not Directus' Right? 😅

"Impossible" is of just a lack of imagination ;) Currently CSP rules are indeed loaded once on startup so making them dynamic is tricky, you'll probably need to inject a new Express Middleware using a custom hook that sets dynamic CSP headers on-request.

**Comment by xgenvn** [comment]
Personally I think in a production environment, try to apply other proxy/api gateway that can add the headers, or this is what I'm currently applying. Exposing Directus directly will be somewhat non-safety setup.

**Comment by br41nslug** [comment]
I am moving this to a Feature Request discussion as this will take a little more detail/planning/discussion before it can be implemented in core
